src/lib/save_load.js

```diff
@@ -0,0 +1,58 @@
+export const load = {
+  readFile(file) {
+    return new Promise((resolve, reject) => {
+      const reader = new FileReader();
+      reader.onload = () => {
+        resolve(reader.result);
+      };
+      reader.onerror = reject;
+      reader.readAsText(file);
+    });
+  },
+
+  async loadJSONLogs(logFile) {
+    const logStr = await this.readFile(logFile);
+    return JSON.parse(logStr);
+  },
+};
+
+export const save = {
+  async downloadFile(blob, filename) {
+    const url = URL.createObjectURL(blob);
+    let downloadId = null;
+    let listener;
+
+    const downloadDonePromise = new Promise((resolve, reject) => {
+      listener = (result) => {
+        if (result.state.current === 'complete' && result.id === downloadId) {
+          resolve(result.state.current);
+        }
+        if (result.error) {
+          reject(result.error);
+        }
+      };
+
+      browser.downloads.onChanged.addListener(listener);
+    });
+
+    try {
+      downloadId = await browser.downloads.download({
+        url,
+        filename,
+      });
+      return await downloadDonePromise;
+    } catch (error) {
+      return error;
+    } finally {
+      URL.revokeObjectURL(url);
+      browser.downloads.onChanged.removeListener(listener);
+    }
+  },
+
+  async saveAsJSON(logs) {
+    const blob = new Blob([JSON.stringify(logs)], {
+      type: 'application/json',
+    });
+    return await this.downloadFile(blob, 'activitylogs.json');
```

Comment:
Note: The difference between

```
return await promise;
```

and

```
return promise;
```

... is that in the first case, your function will await the promise and be able to catch errors (which is why it is important to use `return await` in the `downloadFile` method),
whereas in the second case the function returns a value (the promise) and any promise rejections will directly be sent to the caller. In the `saveAsJSON` method, it is not necessary to use `await` in `return await ...` because the promise result is not handled but directly passed to the caller.